File: bindings/python/tests/test_flx_cuda.py
"""test_flx_cuda.py — Python port test suite for flx-cuda CPU fallback.

Runs without CUDA. Tests the same invariants as the C++ test suite.

Run with: python3 test_flx_cuda.py
"""

import sys
import os

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from flx_cuda import OpCell, PriorityScheduler, QuiltSchedulerView


def test_submit_and_dispatch():
    """Submit 3 ops at different priorities; dispatch should return highest first."""
    sched = PriorityScheduler(use_cuda=False)
    sched.submit(OpCell(kind=0, priority=5, size=4096))
    sched.submit(OpCell(kind=1, priority=9, ptr=0x1000))
    sched.submit(OpCell(kind=2, priority=3, ptr=0x2000))

    op = sched.dispatch()
    assert op is not None, "expected first dispatch to succeed"
    assert op.priority == 9, f"expected priority 9, got {op.priority}"
    sched.complete(op.id)

    op = sched.dispatch()
    assert op.priority == 5, f"expected priority 5, got {op.priority}"
    sched.complete(op.id)

    op = sched.dispatch()
    assert op.priority == 3, f"expected priority 3, got {op.priority}"
    sched.complete(op.id)

    op = sched.dispatch()
    assert op is None, "expected empty queue"

    print("test_submit_and_dispatch: ok")


def test_drain():
    """Drain should empty a bucket."""
    sched = PriorityScheduler(use_cuda=False)
    for i in range(5):
        sched.submit(OpCell(kind=0, priority=5, size=1024 * (i + 1)))

    counts = sched.queue_counts()
    assert counts[5 - sched.min_priority] == 5

    removed = sched.drain(5)
    assert removed == 5, f"expected 5 removed, got {removed}"

    counts = sched.queue_counts()
    assert counts[5 - sched.min_priority] == 0

    print("test_drain: ok")


def test_quilt_projection():
    """The 5+1 Quilt opcodes applied to the scheduler."""
    sched = PriorityScheduler(use_cuda=False)
    view = QuiltSchedulerView(sched)

    # BIND
    view.bind()

    # EFFECT submit
    view.effect_submit(OpCell(kind=0, priority=5, size=8192))
    view.effect_submit(OpCell(kind=1, priority=7, ptr=0x3000))

    # VIEW
    counts = view.view_counts()
    assert counts[5 - sched.min_priority] == 1
    assert counts[7 - sched.min_priority] == 1

    # TICK
    view.tick(dt=1)
    assert sched.tick_count == 1

    # EFFECT dispatch (highest first)
    op = view.effect_dispatch()
    assert op.priority == 7, f"expected priority 7, got {op.priority}"

    # EFFECT complete
    view.effect_complete(op.id)

    # Next dispatch
    op = view.effect_dispatch()
    assert op.priority == 5

    # FORGET
    view.forget(priority=5)

    # Witness chain
    assert len(view.witness) >= 7, f"expected ≥7 witness events, got {len(view.witness)}"

    print("test_quilt_projection: ok")


def test_priority_out_of_range():
    """Submitting with priority out of [min, max] should still work (bucket lookup)."""
    sched = PriorityScheduler(use_cuda=False, min_priority=0, max_priority=9)
    sched.submit(OpCell(kind=0, priority=3, size=4096))
    op = sched.dispatch()
    assert op.priority == 3
    print("test_priority_out_of_range: ok")


def main():
    failed = 0
    for test in [test_submit_and_dispatch, test_drain, test_quilt_projection, test_priority_out_of_range]:
        try:
            test()
        except AssertionError as e:
            print(f"FAIL {test.__name__}: {e}")
            failed += 1
        except Exception as e:
            print(f"ERROR {test.__name__}: {e}")
            failed += 1

    if failed == 0:
        print(f"\nAll {len([test_submit_and_dispatch, test_drain, test_quilt_projection, test_priority_out_of_range])} flx-cuda Python tests passed.")
        return 0
    print(f"\n{failed} test(s) failed.")
    return 1


if __name__ == '__main__':
    sys.exit(main())

File: bindings/rust/tests/test_flx_cuda.rs
//! flx-cuda Rust test suite — same invariants as C++/Python test suites.
//!
//! Run with: cargo test
//!
//! Author: SuperInstance
//! License: MIT

use flx_cuda::{Error, OpCell, OpStatus, QuiltView, Scheduler};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn make_op(id: i32, kind: i8, priority: i8, size: i32, ptr: u64) -> OpCell {
    OpCell {
        id,
        kind,
        priority,
        status: OpStatus::Queued as i8,
        _pad: 0,
        size,
        ptr,
        timestamp: 0,
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[test]
fn test_init_and_bind() {
    let mut sched = Scheduler::new(1024, -128, 127);
    sched.bind();
    assert_eq!(sched.tick_count(), 0);
    assert_eq!(sched.total_ops(), 0);
    assert!(!sched.witness().is_empty(), "BIND event should be recorded");
    assert!(sched.witness()[0].contains("BIND"));
}

#[test]
fn test_submit_and_dispatch_priority_order() {
    let mut sched = Scheduler::new(1024, 0, 9);
    sched.bind();

    // Submit at priorities 5, 9, 3
    sched.submit(make_op(1, 0, 5, 4096, 0)).unwrap();
    sched.submit(make_op(2, 1, 9, 0, 0x1000)).unwrap();
    sched.submit(make_op(3, 2, 3, 0, 0x2000)).unwrap();

    // Dispatch should yield priority 9 first
    let op = sched.dispatch().unwrap();
    assert_eq!(op.id, 2);
    assert_eq!(op.priority, 9);
    sched.complete(op.id).unwrap();

    let op = sched.dispatch().unwrap();
    assert_eq!(op.id, 1);
    assert_eq!(op.priority, 5);
    sched.complete(op.id).unwrap();

    let op = sched.dispatch().unwrap();
    assert_eq!(op.id, 3);
    assert_eq!(op.priority, 3);
    sched.complete(op.id).unwrap();

    // Queue empty
    assert_eq!(sched.dispatch().unwrap_err(), Error::QueueEmpty);
}

#[test]
fn test_drain() {
    // Capacity must be enough: 256 buckets × 5 ops/bucket = 1280 minimum.
    let mut sched = Scheduler::new(2048, -128, 127);
    sched.bind();

    for i in 0..5 {
        sched.submit(make_op(100 + i, 0, 5, 1024 * (i + 1), 0)).unwrap();
    }
    let counts = sched.view_counts();
    assert_eq!(counts[133], 5);  // priority 5 → bucket 133

    let removed = sched.drain(5).unwrap();
    assert_eq!(removed, 5);

    let counts = sched.view_counts();
    assert_eq!(counts[133], 0);
}

#[test]
fn test_priority_out_of_range() {
    let mut sched = Scheduler::new(1024, 0, 9);
    sched.bind();
    let r = sched.submit(make_op(1, 0, 100, 1024, 0));
    assert_eq!(r.unwrap_err(), Error::InvalidPriority);
}

#[test]
fn test_quilt_projection_invariants() {
    // The 5+1 Quilt opcodes applied via the QuiltView
    let mut sched = Scheduler::new(1024, 0, 9);
    let mut view = QuiltView::new(&mut sched);

    view.bind();
    view.effect_submit(make_op(1, 0, 5, 8192, 0)).unwrap();
    view.effect_submit(make_op(2, 1, 7, 0, 0x3000)).unwrap();

    let counts = view.view_counts();
    assert_eq!(counts[5], 1);
    assert_eq!(counts[7], 1);

    view.tick(1);
    assert_eq!(view.scheduler.tick_count(), 1);

    let op = view.effect_dispatch().unwrap();
    assert_eq!(op.priority, 7);

    view.effect_complete(op.id).unwrap();

    let op = view.effect_dispatch().unwrap();
    assert_eq!(op.priority, 5);

    view.forget(5).unwrap();

    // Witness chain should be non-empty and contain key events
    let witness = view.witness();
    let has_bind = witness.iter().any(|w| w.contains("BIND"));
    let has_submit = witness.iter().any(|w| w.contains("submit"));
    let has_dispatch = witness.iter().any(|w| w.contains("dispatch"));
    let has_complete = witness.iter().any(|w| w.contains("complete"));
    let has_drain = witness.iter().any(|w| w.contains("FORGET"));

    assert!(has_bind, "BIND event missing");
    assert!(has_submit, "submit event missing");
    assert!(has_dispatch, "dispatch event missing");
    assert!(has_complete, "complete event missing");
    assert!(has_drain, "FORGET event missing");
}

#[test]
fn test_priority_ordering_under_load() {
    // Submit 100 ops at random priorities; dispatch yields them
    // in priority-descending order.
    let mut sched = Scheduler::new(1024, -128, 127);
    sched.bind();

    for i in 0..100 {
        // Deterministic priority sequence in range [-100, 100] (no overflow)
        let priority = (((i * 37) % 200) - 100) as i8;
        sched.submit(make_op(i + 1, 0, priority, 1024, 0)).unwrap();
    }

    let mut prev = i8::MAX;
    let mut count = 0;
    while let Ok(op) = sched.dispatch() {
        assert!(op.priority <= prev, "priority must be non-increasing");
        prev = op.priority;
        count += 1;
        sched.complete(op.id).unwrap();
    }
    assert_eq!(count, 100);
}

#[test]
fn test_auto_id_assignment() {
    let mut sched = Scheduler::new(1024, 0, 9);
    sched.bind();
    sched.submit(make_op(0, 0, 5, 1024, 0)).unwrap();
    sched.submit(make_op(0, 0, 5, 1024, 0)).unwrap();
    let op = sched.dispatch().unwrap();
    assert_eq!(op.id, 1);
    let op = sched.dispatch().unwrap();
    assert_eq!(op.id, 2);
}
